catalyst-toolbox-lib/src/recovery/tally.rs

```diff
@@ -0,0 +1,648 @@
+            let PersistentFragmentLog { fragment, time } = persistent_fragment_log;
+            match fragment {
+                Fragment::VoteCast(ref transaction) => {
+                    let transaction_slice = transaction.as_slice();
+                    let is_valid_vote_cast = valid_vote_cast(&transaction_slice);
+                    if !is_valid_vote_cast {
+                        return Some(Err((fragment, ValidationError::InvalidVoteCast)));
+                    }
+                    let (_, identifier, witness) =
+                        deconstruct_account_transaction(&transaction_slice);
+                    if !self.validate_timestamp(time, &self.vote_start, &self.vote_end) {
+                        return Some(Err((fragment, ValidationError::VotingPeriodError)));
+                    }
+                    if !self.validate_tx(&identifier, &witness, &transaction_slice) {
+                        return Some(Err((
+                            fragment.clone(),
+                            ValidationError::InvalidTransactionSignature {
+                                id: fragment.clone().hash().to_string(),
+                                range: self.range_check.clone(),
+                            },
+                        )));
+                    }
+
+                    // check if fragment was processed already
+                    let fragment_id = fragment.id();
+                    if self.replay_protection.contains(&fragment_id) {
+                        return Some(Err((
+                            fragment,
+                            ValidationError::DuplicatedFragment { id: fragment_id },
+                        )));
+                    }
+
+                    self.replay_protection.insert(fragment_id);
+
+                    *self.spending_counters.get_mut(&identifier).unwrap() += 1;
+                    Some(Ok(fragment))
+                }
+                Fragment::VoteTally(ref transaction) => {
+                    let (_, identifier, witness) =
+                        deconstruct_account_transaction(&transaction.as_slice());
+                    if !self.validate_timestamp(time, &self.vote_end, &self.committee_end) {
+                        return Some(Err((fragment, ValidationError::TallyPeriodError)));
+                    }
+                    if !self.validate_tx(&identifier, &witness, &transaction.as_slice()) {
+                        return Some(Err((
+                            fragment.clone(),
+                            ValidationError::InvalidTransactionSignature {
+                                id: fragment.hash().to_string(),
+                                range: self.range_check.clone(),
+                            },
+                        )));
+                    }
+                    *self.spending_counters.get_mut(&identifier).unwrap() += 1;
+                    Some(Ok(fragment))
+                }
+                _ => None,
```

Comment:
This actually will make the `for fragment in ....` stop if there's a fragment that's not a vote cast or tally, we should skip that instead

Reply:
Yes, the tool should not silently drop the rest of the chain.

Reply:
I leveraged this as an "error", mostly so the fragment will be discarded and reported with the other fragments. I think it would be good for forensics.